File: Data Structures & Algorithms/contains-duplicate-ii/submission-3.py
class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        hm = {}
        for i, num in enumerate(nums):
            if num in hm and abs(hm[num] - i) <= k:
                return True
            hm[num] = i
        return False

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-3.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int total = ranges::fold_left(nums, 0, plus<>{});
        if (total % 2 != 0) return false;

        int half_total = total / 2;
        vector<bool> dp(half_total + 1, false);
        dp[0] = true;

        for (int num : nums) {
            for (int sum = half_total; sum >= num; sum--) {
                dp[sum] = dp[sum] || dp[sum - num];
            }
        }
        return dp[half_total];
    }
};
